File: lxrbckl/local.py
# import <
from json.decoder import JSONDecodeError
from json import (
   
   dump,
   load
   
)
from os import (
   
   getcwd,
   remove
   
)
from os.path import (
   
   isfile,
   dirname,
   basename
   
)

# >


def getProjectPath(
   
   pFile: str = '',
   pDelimeter: str = '/'
   
):
   '''  '''
   
   cwd = getcwd()
   path = [dirname(cwd), basename(cwd), pFile]
   
   return pDelimeter.join(path)
   

def fileSet(
   
   pData,
   pFile: str,
   pIndent: int = 3,
   pOverride: bool = True,
   pPath: str = getProjectPath()
   
):
   '''  '''
      
   # try (if permitted) <
   # except (then not permitted) <
   try:

      # if (write to file) <
      # else (throw error) <
      if ((isfile(f'{pPath}{pFile}') is False) or (pOverride)):
         
         with open(pFile, 'w') as fout:
            
            {
               
               'txt' : lambda : fout.write(pData),
               'json' : lambda : dump(
                  
                  fp = fout,
                  obj = pData,
                  indent = pIndent
                  
               )
               
            }[pFile.split('.')[1]]()

      else: return 'File already exists.'

      # >

   except TypeError: return 'Invalid data for this filetype.'
   
   # >


def fileGet(
   
   pFile: str,
   pPath: str = getProjectPath()
   
):
   '''  '''
   
   # try (if existing file) <
   # except (then bad .json format) <
   # except (then non-existing file) <
   try:
   
      with open(pFile, 'r') as fin:
         
         return {
            
            'txt' : lambda : fin.read(),
            'json' : lambda : load(fin)
            
         }[pFile.split('.')[1]]()
   
   except JSONDecodeError: return 'Loaded data is broken.'
   except FileNotFoundError: return 'File does not exist.'
   
   # >


def fileDel(
   
   pFile: str,
   pPath: str = getProjectPath()

):
   '''  '''
   
   # try (if existing file) <
   # except (then non-existing file) <
   try: remove(f'{pPath}{pFile}')
   except FileNotFoundError: return 'File does not exist.'
   
   # >

File: local.js
// import <
const fs = require('fs').promises;
const path = require('path');

// >


function getProjectPath(
   
   pFile = '',
   pDelimeter = '/'
   
) {

   const dir = path.dirname(__filename);
   const file = pFile.split(/[/\\]/).join(pDelimeter);

   return [dir, file].join(pDelimeter);

}


async function fileSet(

   pData,
   pFile,
   pPath = getProjectPath()

) {

   try {

      fs.writeFile(

         file = pPath + pFile,
         data = {

            'txt' : () => {return pData;},
            'json' : () => {return JSON.stringify(pData);}

         }[pFile.split('.').slice(-1)]()

      );

   } catch (error) {return 'Could not write to file.';}

}


async function fileGet(

   pFile,
   pEncoding = 'utf8',
   pPath = getProjectPath()

) {

   try {

      const fin = await fs.readFile(

         file = pPath + pFile,
         options = {encoding : pEncoding}

      );

      return {

         'txt' : () => {return fin;},
         'json' : () => {return JSON.parse(fin);}

      }[pFile.split('.').slice(-1)]();

   } catch (error) {return 'Could not read file.';}

}


(async () => {

   // set

   // get
   const file = 'sj.json';
   await fileSet(

      pData = {"test" : "test"},
      pFile = file

   );
   const fin = await fileGet(pFile = file);
   console.log(fin);

})();


async function fileDel(

   pFile,
   pPath = getProjectPath()

) {



}


// export <
module.exports = {

   fileSet,
   fileGet,
   fileDel

};

// >